viewing_party/party.py

```diff
@@ -1,23 +1,179 @@
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie["title"])
+    friend_movies = set(friend_movies)
+    
+    for my_watched in user_data["watched"]:
+        if my_watched["title"] not in friend_movies:
+            my_movies.append(my_watched)
+
+    return my_movies
+
+def get_friends_unique_watched(user_data):
+
+    friend_movies = []
+    my_movies = []
+
+    for my_watched in user_data["watched"]:
+        my_movies.append(my_watched["title"])
+    my_movies = set(my_movies)
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie not in friend_movies: 
+                if movie["title"] not in my_movies:
+                    friend_movies.append(movie)
+    
+    return friend_movies
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+
+    recommended_movies = []
+    movies_user_watched = []
+    
+    if not user_data["watched"]:
+        return recommended_movies
+
+    for elem in user_data["watched"]:
+        movies_user_watched.append(elem["title"])
+    
+    for friends in user_data["friends"]:
+        for movie in friends["watched"]:
+            if movie["title"] not in movies_user_watched:
+                recommended_movies.append(movie)
+    
+    for recommendation in recommended_movies:
+        if recommendation["host"] not in user_data["subscriptions"]:
+            recommended_movies.remove(recommendation)
+    
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    favorite_genres = []
+    recommended_movies = get_available_recs(user_data)
+    
+    if not recommended_movies:
+        return recommended_movies
+
+    for elem in user_data["watched"]:
+        favorite_genres.append(elem["genre"])
+        
+    counters = Counter(favorite_genres)
+    favorite_genre = max(counters, key=counters.get)
+
+    for recommendation in recommended_movies:
+        if recommendation["genre"] != favorite_genre:
+            recommended_movies.remove(recommendation)
+    
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    recommended_movies = []
+    friend_movies = []
+    
+    if not user_data["watched"] or not user_data["friends"]:
+        return recommended_movies
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+
+    for movie in user_data["watched"]:
+        if movie not in friend_movies:
+            recommended_movies.append(movie)
+
+    for recommendation in recommended_movies:
+        if recommendation not in user_data["favorites"]:
+            recommended_movies.remove(recommendation)

```

Comment:
A way to remove this last for loop would be to add an additional condition to line 172 so you don't need to add movies and then remove them in another loop.

Line 172 would look like: 
`if movie not in friend_movies and movie in user_data["favorites"]:`

then you could delete lines 175-177